openstudiocore/src/model/ExternalFile.cpp

```diff
@@ -0,0 +1,188 @@
+                                       Model_Impl* model,
+                                       bool keepHandle)
+    : ResourceObject_Impl(other,model,keepHandle)
+  {}
+
+  const std::vector<std::string>& ExternalFile_Impl::outputVariableNames() const
+  {
+    static std::vector<std::string> result;
+    if (result.empty()){
+    }
+    return result;
+  }
+
+  IddObjectType ExternalFile_Impl::iddObjectType() const {
+    return ExternalFile::iddObjectType();
+  }
+
+  std::string ExternalFile_Impl::fileName() const {
+    boost::optional<std::string> value = getString(OS_External_FileFields::FileName,true);
+    OS_ASSERT(value);
+    return value.get();
+  }
+
+  boost::optional<std::string> ExternalFile_Impl::columnSeparator() const {
+    return getString(OS_External_FileFields::ColumnSeparator,true);
+  }
+
+  char ExternalFile_Impl::columnSeparatorChar() const {
+    static std::unordered_map<std::string, char> lookup({ { "Comma", ',' }, { "Tab", '\t' }, { "Fixed", ' ' }, { "Semicolon", ';' } });
+    boost::optional<std::string> value = getString(OS_External_FileFields::ColumnSeparator, true);
+    OS_ASSERT(value);
+    auto it = lookup.find(value.get());
+    if (it == std::end(lookup)) {
+      // Invalid separator
+      return '\0';
+    }
+    return it->second;
+  }
+
+  bool ExternalFile_Impl::isColumnSeparatorDefaulted() const {
+    return isEmpty(OS_External_FileFields::ColumnSeparator);
+  }
+
+  bool ExternalFile_Impl::setFileName(const std::string& fileName) {
+    bool result = setString(OS_External_FileFields::FileName, fileName);
+    OS_ASSERT(result);
+    return result;
+  }
+
+  bool ExternalFile_Impl::setColumnSeparator(const std::string& columnSeparator) {
+    bool result = setString(OS_External_FileFields::ColumnSeparator, columnSeparator);
+    return result;
+  }
+
+  void ExternalFile_Impl::resetColumnSeparator() {
+    bool result = setString(OS_External_FileFields::ColumnSeparator, "");
+    OS_ASSERT(result);
+  }
+
+} // detail
+
+ExternalFile::ExternalFile(const Model& model, const std::string &filename)
+  : ResourceObject(ExternalFile::iddObjectType(),model)
+{
+  OS_ASSERT(getImpl<detail::ExternalFile_Impl>());
+
+  std::ifstream ifs;
+  ifs.open(filename, std::ifstream::in);
+  if (!ifs.is_open()) {
+    // Failed to open file
+    LOG_AND_THROW("Failed to open file \"" << filename << "\" for " << this->briefDescription());
+  }
+  ifs.close();
+
+  bool ok;
+  ok = setFileName(filename);
+  OS_ASSERT(ok);
+
+  // Moving the file etc. should go here
+}
+
+IddObjectType ExternalFile::iddObjectType() {
+  return IddObjectType(IddObjectType::OS_External_File);
+}
+
+std::vector<std::string> ExternalFile::columnSeparatorValues() {
+  return getIddKeyNames(IddFactory::instance().getObject(iddObjectType()).get(),
+                        OS_External_FileFields::ColumnSeparator);
+}
+
+std::string ExternalFile::fileName() const {
+  return getImpl<detail::ExternalFile_Impl>()->fileName();
+}
+
```

Comment:
Could add a helper method here to get the full path to the file, something like:

model().workflowJSON().findFile(fileName());